libphobos/libdruntime/gcc/deh.d

```diff
@@ -26,776 +26,927 @@ import gcc.config;
+}
 
-// Called when fibers switch contexts.
-extern(C) void *
-_d_eh_swapContext(void* newContext) nothrow
+/**
+ * Called when fibers switch contexts.
+ */
+extern(C) void* _d_eh_swapContext(void* newContext) nothrow
 {
-    auto old = __globalExceptions.thrownExceptions;
-    __globalExceptions.thrownExceptions = cast(d_exception_header *) newContext;
+    auto old = ExceptionHeader.stack;
+    ExceptionHeader.stack = cast(ExceptionHeader*)newContext;
     return old;
 }
 
+/**
+ * Called before starting a catch.  Returns the exception object.
+ */
+extern(C) void* __gdc_begin_catch(_Unwind_Exception* unwindHeader)
+{
+    ExceptionHeader* header = ExceptionHeader.toExceptionHeader(unwindHeader);
+
+    void* objectp = cast(void*)header.object;
+
+    // Something went wrong when stacking up chained headers...
+    if (header != ExceptionHeader.pop())
+        terminate(__LINE__);
 
-// Called before starting a catch.  Returns the exception object.
+    // Handling for this exception is complete.
+    _Unwind_DeleteException(&header.unwindHeader);
+
+    return objectp;
+}
 
-extern(C) void *
-__gdc_begin_catch(void *exc_ptr)
+/**
+ * Perform a throw, D style. Throw will unwind through this call,
+ * so there better not be any handlers or exception thrown here.
+ */
+extern(C) void _d_throw(Throwable object)
 {
-  _Unwind_Exception *exceptionObject = cast(_Unwind_Exception *) exc_ptr;
-  d_exception_header *header = get_exception_header_from_ue(exceptionObject);
-  globalExceptions *globals = &__globalExceptions;
+    // If possible, avoid always allocating new memory for exception headers.
+    ExceptionHeader *eh = ExceptionHeader.create(object);
 
-  // Something went wrong when stacking up chained headers...
-  if (header != globals.thrownExceptions)
-    __gdc_terminate();
+    // Add to thrown exception stack.
+    eh.push();
 
-  void *objectp = cast(void *) header.object;
+    // Called by unwinder when exception object needs destruction by other than our code.
+    extern(C) void exception_cleanup(_Unwind_Reason_Code code, _Unwind_Exception* exc)
+    {
+        // If we haven't been caught by a foreign handler, then this is
+        // some sort of unwind error.  In that case just die immediately.
+        // _Unwind_DeleteException in the HP-UX IA64 libunwind library
+        //  returns _URC_NO_REASON and not _URC_FOREIGN_EXCEPTION_CAUGHT
+        // like the GCC _Unwind_DeleteException function does.
+        if (code != _URC_FOREIGN_EXCEPTION_CAUGHT && code != _URC_NO_REASON)
+            terminate(__LINE__);
+
+        auto eh = ExceptionHeader.toExceptionHeader(exc);
+        ExceptionHeader.free(eh);
+    }
+
+    eh.unwindHeader.exception_cleanup = &exception_cleanup;
 
-  // Handling for this exception is complete.
-  globals.thrownExceptions = header.nextException;
-  _Unwind_DeleteException (&header.unwindHeader);
+    // Runtime now expects us to do this first before unwinding.
+    _d_createTrace(eh.object, null);
```

Comment:
That should be the case, yes, as there are no restraints applied to the TraceHandler callback.  Unless the trace handle is explicitly set by user, however, then this is always a noop.